File: backend/src/routers/schemas/movie.py
from typing import List, Optional
from pydantic import BaseModel, Field
from src.models import MovieCreate, MovieUpdate, MovieRead


class MovieCreateRequestSchema(MovieCreate):
    """创建电影请求模型"""
    pass


class MovieUpdateRequestSchema(MovieUpdate):
    """更新电影请求模型"""

    pass  # 所有字段从 MovieUpdate 继承


class MovieReadResponseSchema(MovieRead):
    is_favoriter: Optional[bool] = Field(None, description="当前用户是否收藏了该影片")
    is_watchLater: Optional[bool] = Field(None, description="当前用户是否加入了稍后观看")


class MoviePageResultResponseSchema(BaseModel):
    items: List[MovieReadResponseSchema] = Field(default_factory=list, description="电影列表")
    total: int = Field(0, description="总数量")
    page: int = Field(1, description="当前页码")
    size: int = Field(20, description="每页大小")
    pages: int = Field(0, description="总页数")

# 新增：批量更新请求模型
class MovieBatchUpdateRequestSchema(BaseModel):
    movie_ids: List[str] = Field(..., description="影片ID列表")
    patch: MovieUpdateRequestSchema = Field(..., description="更新内容")


class MovieCreateResponseSchema(BaseModel):
    movie_info: MovieReadResponseSchema = Field(..., description="创建的电影信息")
    task_id: str = Field(..., description="异步任务ID")

File: frontend/src/types/movie.ts
// 电影类型定义，映射后端 models/movies/movie_models.py 与 routers/schemas/movie.py

export interface MovieMetadata {
  duration?: number;
  country: string[];
  language?: string;
}

export interface MovieBase {
  library_id: string;
  title: string;
  title_cn: string;
  directors: string[];
  actors: string[];
  description: string;
  description_cn: string;
  // 后端为 Optional[date]，序列化为字符串或 null
  release_date?: string | null;
  genres: string[];
  // 后端有 default_factory，因此在实际对象中始终存在
  metadata: MovieMetadata;
  // 后端为 Optional[float]，取值范围 0~10
  rating?: number | null;
  tags: string[];
}

export type MovieCreate = MovieBase

export interface MovieUpdate {
  library_id: string;
  title?: string;
  title_cn?: string;
  directors?: string[];
  actors?: string[];
  description?: string;
  description_cn?: string;
  release_date?: string | null;
  genres?: string[];
  metadata?: MovieMetadata;
  rating?: number | null;
  tags?: string[];
}

export interface MovieInDB extends MovieBase {
  id: string;
  has_poster: boolean;
  has_backdrop: boolean;
  has_thumbnail: boolean;
  is_deleted: boolean;
  deleted_at?: string | null;
  created_at: string;
  updated_at: string;
  is_favoriter?: boolean;
  is_watchLater?: boolean;
}

export type MovieRead = MovieInDB

export interface MoviePageResult {
  items: MovieRead[];
  total: number;
  page: number;
  size: number;
  pages: number;
}

// 路由层 Schema（routers/schemas/movie.py）
export type MovieCreateRequestSchema = MovieCreate;

export type MovieUpdateRequestSchema = MovieUpdate

export type MovieReadResponseSchema = MovieRead

export type MoviePageResultResponseSchema = MoviePageResult

export interface MovieCreateResponse {
  movie_info: MovieReadResponseSchema;
  task_id?: string | null;
}
